Add lazy=True keyword argument to qml.s_prod, qml.prod, and qml.op_sum

kevinkayyy opened the pull request:
**Context:** Currently, qml.pow and qml.adjoint both have a keyword argument lazy=True that indicates whether or not to perform some basic simplification steps. We are now doing similar things for qml.s_prod, qml.prod, and qml.op_sum, if lazy=False.

**Description of the Change:** 

For qml.s_prod(scalar, op, lazy=False), we check if op is already a scalar product SProd. If so, we combine the coefficients.
For qml.prod(*ops, lazy=False), we check if any of the provided operators are already of type Prod. If so, we use its operands instead.
For qml.op_sum(*summands, lazy=False), we check if any of the summands are of type Sum. If so, we use its constituents instead. 

**Related GitHub Issues:**

Closes #3094

kevinkayyy commented:
Hi all, 

It seems that all checks have passed. However, please still check if the unit tests added are good enough.

albi3ro commented on pennylane/ops/op_math/prod.py:
```diff
@@ -68,6 +69,12 @@ def prod(*ops, do_queue=True, id=None):
            [ 1,  0]])
     """
+    if not lazy:
+        ops_simp = tuple()
```
Since `ops_simp` is going to be dynamically updated a lot, lists are going to be more efficient than tuples.  Every time we modify the tuple, we would have to create a new object.  If we perform this with a list, python just modifies the existing object.

I don't know if the performance difference will show up for a function like this in practice, but better safe than sorry.  While lists are often overused and abused, this is a case we want mutability.

This can easily be checked running `%%timeit` on both
```
a = tuple()
for i in range(100):
    a += (i,)
```
and 
```
a = []
for i in range(100):
    a += [i]
```

kevinkayyy replied:
Hi Christina, I've replaced tuples by lists, please check.

albi3ro reviewed (commented):
One thing I forgot to mention in the Issue: queuing.

Suppose we did something like this in a quantum function:

```
def qfunc():
    prod1 = qml.prod(qml.S(0), qml.T(0))
    prod2 = qml.prod(prod2, qml.S(1), lazy=False)
```

We would only want `prod2` to end up in the circuit.  `prod1` would only be an intermediary step used to construct `prod2`.

So how do we remove `prod1` from the circuit? You can see an example in the `qml.pow` constructor here: https://github.com/PennyLaneAI/pennylane/blob/bf36a0a007e5d3dfe3ccee2b71787c56c5de5914/pennylane/ops/op_math/pow.py#L92

We give it "metadata" saying that is has an "owner":
```
        QueuingManager.update_info(base, owner=pow_op)
```

We would need to do similar logic for each operation like `prod1` that we need to remove from the circuit.

For example, we would need:
```pycon
>>> with qml.queuing.AnnotatedQueue() as q:
...     prod1 = qml.prod(qml.S(0), qml.T(0))
...     prod2 = qml.prod(prod2, qml.S(1), lazy=False)
>>> q.get_info(prod1)
{'owner': S(wires=[0]) @ T(wires=[0]) @ S(wires=[1])}
```

For more context, we now have developer documentation [on the queuing module](https://docs.pennylane.ai/en/latest/code/qml_queuing.html)

kevinkayyy commented:
Hi Christina, queuing is now added to qml.prod. I looked deeper into the corresponding classes and also their parent classes and I hope my changes are written in a coherent style. 

I also added queuing for qml.op_sum and qml.s_prod but I'm not really sure about those, because I saw in the "_queue_category" functions, there are comments saying "# don't queue scalar prods (Sum instances) as they might not be Unitary!" Please let me know if I should remove queuing for these wrappers, or write something for the classes to make all operators unitary.

albi3ro commented on doc/releases/changelog-dev.md:
```diff
@@ -290,9 +290,13 @@
   the optional `use_grouping` attribute.
   [(#3456)](https://github.com/PennyLaneAI/pennylane/pull/3456)
 
+* Add the optional argument lazy=True to functions qml.s_prod, qml.prod and qml.op_sum to allow simplification.
```
```suggestion
* Add the optional argument lazy=True to functions `qml.s_prod`, `qml.prod` and `qml.op_sum` to allow simplification.
```

albi3ro commented on pennylane/ops/op_math/prod.py:
```diff
@@ -68,7 +70,21 @@ def prod(*ops, do_queue=True, id=None):
+        return Prod(*ops, do_queue=do_queue, id=id)
+
+    ops_simp = Prod(
+        *itertools.chain.from_iterable([op if isinstance(op, Prod) else [op] for op in ops]),
```
+1 great solution :)

kevinkayyy commented:
> Exactly! Looks great 👍
> 
> Not too much to comment on. Exactly what I wanted, and I'm glad to finally have this feature.
> 
> Nice solution using `itertools` and adjustment to update the queuing behaviour. Nice clean tests.

Thanks. I'm glad that you like it.
